File: apply_migration.py
#!/usr/bin/env python3
"""
Apply drill system migration to production Supabase database.
"""
import os
from dotenv import load_dotenv
from supabase import create_client

# Load environment variables
load_dotenv('.env.local')

SUPABASE_URL = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
SUPABASE_SERVICE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY')

if not SUPABASE_URL or not SUPABASE_SERVICE_KEY:
    raise ValueError("Missing Supabase credentials in .env.local")

print(f"🔗 Connecting to Supabase: {SUPABASE_URL}")

# Create Supabase client
supabase = create_client(SUPABASE_URL, SUPABASE_SERVICE_KEY)

# Read migration SQL
print("📖 Reading migration file...")
with open('supabase/migrations/20251220_create_drill_system.sql', 'r') as f:
    migration_sql = f.read()

# Execute migration using RPC or direct SQL
print("🚀 Applying migration to production database...")

try:
    # Use PostgREST to execute raw SQL via the service role
    # Note: We'll need to execute this in chunks to handle the complexity

    # Split by semicolons to execute statements individually
    statements = [s.strip() for s in migration_sql.split(';') if s.strip() and not s.strip().startswith('--')]

    total = len(statements)
    print(f"📊 Executing {total} SQL statements...")

    for i, statement in enumerate(statements, 1):
        # Skip empty statements and comments
        if not statement or statement.startswith('--'):
            continue

        try:
            # Execute via Supabase RPC
            result = supabase.rpc('exec_sql', {'sql': statement}).execute()

            if i % 10 == 0:
                print(f"✓ Progress: {i}/{total} statements executed")

        except Exception as e:
            # Some statements might fail if they already exist (like CREATE TABLE IF NOT EXISTS)
            # Only print errors for non-existence issues
            error_msg = str(e).lower()
            if 'already exists' not in error_msg and 'duplicate' not in error_msg:
                print(f"⚠️  Statement {i} warning: {e}")
                print(f"   Statement: {statement[:100]}...")

    print("✅ Migration applied successfully!")
    print("\n📊 Verifying tables...")

    # Verify tables were created
    verify_query = """
    SELECT table_name,
           (SELECT COUNT(*) FROM information_schema.columns WHERE table_name = t.table_name) AS column_count
    FROM information_schema.tables t
    WHERE t.table_name IN ('drill_scenarios', 'drill_runs', 'drill_executions', 'drill_evaluations')
    AND t.table_schema = 'public'
    ORDER BY t.table_name;
    """

    # Query the tables directly to verify
    scenarios = supabase.table('drill_scenarios').select('id', count='exact').limit(0).execute()
    runs = supabase.table('drill_runs').select('id', count='exact').limit(0).execute()
    executions = supabase.table('drill_executions').select('id', count='exact').limit(0).execute()
    evaluations = supabase.table('drill_evaluations').select('id', count='exact').limit(0).execute()

    print(f"✓ drill_scenarios table: exists (count: {scenarios.count})")
    print(f"✓ drill_runs table: exists (count: {runs.count})")
    print(f"✓ drill_executions table: exists (count: {executions.count})")
    print(f"✓ drill_evaluations table: exists (count: {evaluations.count})")

    print("\n✅ All drill system tables verified!")

except Exception as e:
    print(f"❌ Error applying migration: {e}")
    print("\nTrying alternative approach using psycopg2...")

    # Alternative: Use psycopg2 if available
    try:
        import psycopg2
        from urllib.parse import urlparse

        # Parse Supabase URL to get connection details
        # For Supabase, we need to use the direct database URL
        db_url = os.getenv('DATABASE_URL') or f"{SUPABASE_URL}/rest/v1/"

        print("⚠️  Direct SQL execution requires DATABASE_URL environment variable")
        print("    Please run migration manually via Supabase SQL Editor")
        print(f"    URL: {SUPABASE_URL.replace('https://', 'https://')}/project/_/sql")

    except ImportError:
        print("ℹ️  Psycopg2 not available")
        print("    Please run migration manually via Supabase SQL Editor")
        print(f"    URL: {SUPABASE_URL}/project/_/sql")

File: scripts/apply-migration.mjs
#!/usr/bin/env node

/**
 * Apply drill system migration to production Supabase database
 */

import { createClient } from '@supabase/supabase-js';
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Read environment variables from .env.local
const envContent = readFileSync(join(__dirname, '../.env.local'), 'utf-8');
const env = {};
envContent.split('\n').forEach(line => {
  const match = line.match(/^([^=# ]+)=["']?([^"'\n]*)["']?$/);
  if (match) {
    env[match[1]] = match[2];
  }
});

const SUPABASE_URL = env.SUPABASE_URL;
const SUPABASE_SERVICE_ROLE_KEY = env.SUPABASE_SERVICE_ROLE_KEY;

if (!SUPABASE_URL || !SUPABASE_SERVICE_ROLE_KEY) {
  console.error('❌ Missing Supabase credentials in .env.local');
  process.exit(1);
}

console.log('🔗 Connecting to Supabase...');
console.log(`📍 URL: ${SUPABASE_URL}`);

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY);

// Read migration SQL
const migrationPath = join(__dirname, '../supabase/migrations/20251220_create_drill_system.sql');
console.log(`\n📖 Reading migration: ${migrationPath}`);
const migrationSQL = readFileSync(migrationPath, 'utf-8');

console.log(`📊 Migration size: ${migrationSQL.length} characters`);

// Split SQL into executable statements
const statements = migrationSQL
  .split(';')
  .map(s => s.trim())
  .filter(s => s.length > 0 && !s.startsWith('--'));

console.log(`📝 Found ${statements.length} SQL statements to execute\n`);

// Execute migration using Supabase SQL function
console.log('🚀 Applying migration to production database...\n');

try {
  // For Supabase, we need to use their SQL execution method
  // Unfortunately, there's no direct way to execute raw SQL via the JS client
  // We'll need to use the REST API directly

  const response = await fetch(`${SUPABASE_URL}/rest/v1/rpc/exec_sql`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'apikey': SUPABASE_SERVICE_ROLE_KEY,
      'Authorization': `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`
    },
    body: JSON.stringify({ query: migrationSQL })
  });

  if (!response.ok) {
    // exec_sql RPC doesn't exist, we need a different approach
    console.log('⚠️  Direct SQL execution not available via REST API');
    console.log('📋 Using alternative approach: Manual table creation\n');

    // Alternative: Provide instructions for manual application
    console.log('='.repeat(60));
    console.log('MIGRATION APPLICATION REQUIRED');
    console.log('='.repeat(60));
    console.log('\nThe drill system migration needs to be applied manually.');
    console.log('\nOption 1: Supabase SQL Editor (Recommended)');
    console.log('1. Open: https://supabase.com/dashboard/project/rpkkkbufdwxmjaerbhbn/sql/new');
    console.log('2. Copy and paste the contents of:');
    console.log('   supabase/migrations/20251220_create_drill_system.sql');
    console.log('3. Click "Run" to execute the migration');

    console.log('\nOption 2: Using Supabase CLI');
    console.log('1. Run: supabase link --project-ref rpkkkbufdwxmjaerbhbn');
    console.log('2. Enter your database password');
    console.log('3. Run: supabase db push');

    console.log('\n' + '='.repeat(60) + '\n');

    // Verify current state
    console.log('🔍 Verifying current database state...\n');

    const tables = ['drill_scenarios', 'drill_runs', 'drill_executions', 'drill_evaluations'];
    let existingCount = 0;

    for (const table of tables) {
      try {
        const { count, error } = await supabase
          .from(table)
          .select('id', { count: 'exact', head: true });

        if (!error) {
          console.log(`✓ ${table}: exists (count: ${count || 0})`);
          existingCount++;
        }
      } catch (e) {
        console.log(`✗ ${table}: does not exist`);
      }
    }

    console.log(`\n📊 Status: ${existingCount}/4 drill tables exist`);

    if (existingCount === 0) {
      console.log('\n❗ No drill tables found. Migration needs to be applied.');
      console.log('   Please use one of the options above to apply the migration.\n');
      process.exit(1);
    } else if (existingCount === 4) {
      console.log('\n✅ All drill tables exist! Migration already applied.\n');
      process.exit(0);
    } else {
      console.log('\n⚠️  Partial migration detected. Please complete the migration.\n');
      process.exit(1);
    }
  }

  const result = await response.json();
  console.log('✅ Migration applied successfully!');
  console.log(result);

} catch (error) {
  console.error('❌ Error:', error.message);
  process.exit(1);
}
